config/src/main/java/com/typesafe/config/impl/ConfigDelayedMerge.java

```diff
@@ -129,9 +129,18 @@ else if (end instanceof Unmergeable) {
             if (ConfigImpl.traceSubstitutionsEnabled())
                 ConfigImpl.trace(newContext.depth(), "Resolving highest-priority item in delayed merge " + end
                         + " against " + sourceForEnd + " endWasRemoved=" + (source != sourceForEnd));
-            ResolveResult<? extends AbstractConfigValue> result = newContext.resolve(end, sourceForEnd);
-            AbstractConfigValue resolvedEnd = result.value;
-            newContext = result.context;
+
+
+            AbstractConfigValue resolvedEnd = null;
+            try {
+                ResolveResult<? extends AbstractConfigValue> result = newContext.resolve(end, sourceForEnd);
```

Comment:
The bug as I understand it is that here we're resolving more than we need to in 
order to eliminate the delayed merge. We really only need to resolve one "level" and then merge the values together.
We don't need to resolve recursively.

I came up with another fix in https://github.com/lightbend/config/pull/697 that fixes the immediate bug by avoiding the situation where we have two objects here that can be merged without having to resolve them, but my fix may not handle the case where two substitutions in the delayed merge resolve to objects.

Reply:
That definitely looks like a better solution to me, closing the PR :)